test/test_timers.py

```diff
@@ -1,66 +1,131 @@
+import logging
+import time
+
 import eventlet
-from eventlet import Timeout
-from mock import create_autospec
+import pytest
+from mock import Mock
 
-from nameko.containers import ServiceContainer
+from nameko.testing.services import get_extension
 from nameko.testing.utils import wait_for_call
 from nameko.timer import Timer
 
 
-def spawn_managed_thread(fn, identifier=None):
-    return eventlet.spawn(fn)
-
-
-def test_provider():
-    container = create_autospec(ServiceContainer)
-    container.service_name = "service"
-    container.spawn_managed_thread = spawn_managed_thread
-
-    timer = Timer(interval=0.1).bind(container, "method")
-    timer.setup()
-    timer.start()
-
-    assert timer.interval == 0.1
-
-    with wait_for_call(1, container.spawn_worker) as spawn_worker:
-        with Timeout(1):
-            timer.stop()
-
-    # the timer should have stopped and should only have spawned
-    # a single worker
-    spawn_worker.assert_called_once_with(timer, (), {})
-    assert timer.gt.dead
-
-
-def test_stop_timer_immediately():
-    container = create_autospec(ServiceContainer)
-    container.service_name = "service"
-    container.config = {}
-
-    timer = Timer(interval=5).bind(container, "method")
-    timer.setup()
-    timer.start()
-
-    eventlet.sleep(0.1)
-    timer.stop()
-
-    assert container.spawn_worker.call_count == 0
-    assert timer.gt.dead
-
-
-def test_kill_stops_timer():
-    container = create_autospec(ServiceContainer)
-    container.service_name = "service"
-    container.spawn_managed_thread = spawn_managed_thread
-
-    timer = Timer(interval=0).bind(container, "method")
-    timer.setup()
-    timer.start()
-
-    with wait_for_call(1, container.spawn_worker):
-        timer.kill()
+@pytest.mark.parametrize('timer_only', [False, True])
+@pytest.mark.parametrize('interval,eager,call_duration,expected_calls', [
+    (0.1, False, 0, 1),
+    (0.1, True, 0, 2),
+    (5, False, 0, 0),
+    (5, True, 0, 1),
+    # call duration delays second call
+    (0.1, False, 0.2, 1),  # fires at 0.1
+    (0.1, True, 0.2, 1),  # fires at 0, 0.2
+    (0.025, False, 0.07, 2),  # fires at 0.25, 0.095, 0.165
+    (0.025, True, 0.07, 3),  # fires at 0, 0.07, 0.14
+])
+def test_timer_run(interval, eager, call_duration, expected_calls,
+                   timer_only, container_factory):
+    """Test running the timers main loop.
+
+    We test with "timer_only" mode, where only the main loop is run as well as
+    in a more comprehensive mode where the entire container is tested.
+
+    """
+    timeout = 0.15
+    Service = _get_timer_test_service(interval, eager)
```

Comment:
did this.  A question for future code, so I do this correctly next time:  Is the general philosophy in the tests to keep each test as independent as possible and compromising a bit on "DRY"?  I noticed that the tracker is also defined in every test module instead of in `conftest.py` where it could be used everywhere.

Reply:
That is the general philosophy, yes. Although in this case it's not intentional -- the `tracker` was added for one test and then the pattern spread to others, and now we used in lots of places. It's probably due for a refactor into `conftest`.